agentos/component_run.py

```diff
@@ -0,0 +1,227 @@
+from pathlib import Path
+from typing import Any, Optional
+from mlflow.utils.mlflow_tags import MLFLOW_RUN_NAME
+from agentos.run import Run
+from agentos.run_command import RunCommand
+from agentos.registry import Registry
+from agentos.specs import RunSpec
+from agentos.exceptions import PythonComponentSystemException
+
+
+def active_component_run(
+    caller: Any, fail_if_none: bool = False
+) -> Optional[Run]:
+    """
+    A helper function, returns the currently active ComponentRun, if it exists,
+    else None. More specifically, if the caller is an object that is managed by
+    a Component (i.e. if it has a __component__ attribute) that itself has an
+    active_run, return that Run.
+
+    :param caller: the managed object to fetch the active component run for.
+    :param fail_if_none: if no active component run found, throw an exception
+        instead of returning None.
+    :return: the active component run if it exists, else None.
+    """
+    from agentos.component import Component
+
+    if isinstance(caller, Component):
+        component = caller
+    else:
+        try:
+            component = caller.__component__
+        except AttributeError:
+            raise PythonComponentSystemException(
+                "active_run() was called on an object that is not "
+                "managed by a Component. Specifically, the object passed "
+                "to active_run() must have a ``__component__`` attribute."
+            )
+    if not component.active_run:
+        if fail_if_none:
+            raise PythonComponentSystemException(
+                "active_run() was passed an object managed by a Component "
+                "with no active_run, and fail_if_no_active_run flag was "
+                "True."
+            )
+        else:
+            return None
+    else:
+        return component.active_run
+
+
+class ComponentRun(Run):
+    IS_FROZEN_KEY = "agentos.spec_is_frozen"
+    IS_COMPONENT_RUN_TAG = "pcs.is_component_run"
+    RUN_COMMAND_ID_KEY = "agentos.run_command_id"
+    RUN_COMMAND_REGISTRY_FILENAME = "agentos.run_command_registry.yaml"
+    """
+    A ComponentRun represents the execution of a specific entry point of a
+    specific Component with a specific parameter set.
+    """
+
+    def __init__(
+        self,
+        run_command: RunCommand = None,
+        experiment_id: str = None,
+        existing_run_id: str = None,
+    ) -> None:
+        assert (
+            run_command or existing_run_id
+        ), "One of 'run_command' or 'existing_run_id' must be provided."
+        super().__init__(
+            experiment_id=experiment_id, existing_run_id=existing_run_id
+        )
+        assert not (
+            run_command and existing_run_id
+        ), "`run_command` cannot be passed with `existing_run_id`."
+        self._run_command = None
+        if run_command:
+            self._run_command = run_command
+            self.log_run_command(run_command)
```

Comment:
It seems (maybe) a little odd that we assign `run_command` to `self._run_command` and then log it by passing in the `run_command` as an argument (instead of just using `self._run_command` in `log_run_command()` function).  

Is `log_run_command()` used anywhere else such that it needs the `run_command` argument?

Reply:
Yeah, this is confusing. I moved the assignment into the function and the function `set_and_log_run_command`.